File: backend/src/main/java/com/hostel/backend/controller/StudentProfileController.java
package com.hostel.backend.controller;

import com.hostel.backend.dto.*;
import com.hostel.backend.service.StudentProfileService;
import jakarta.validation.Valid;
import org.springframework.data.domain.Page;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
public class StudentProfileController {

    private final StudentProfileService profileService;

    public StudentProfileController(StudentProfileService profileService) {
        this.profileService = profileService;
    }

    // ── Student endpoints ─────────────────────────────────────────────────────

    @PostMapping("/api/profile")
    public ResponseEntity<StudentProfileResponse> submitProfile(
            @Valid @RequestBody StudentProfileRequest request,
            Authentication auth) {
        return ResponseEntity.ok(profileService.submitProfile(auth.getName(), request));
    }

    @GetMapping("/api/profile")
    public ResponseEntity<StudentProfileResponse> getMyProfile(Authentication auth) {
        return ResponseEntity.ok(profileService.getMyProfile(auth.getName()));
    }

    @GetMapping("/api/profile/status")
    public ResponseEntity<Map<String, Boolean>> checkProfileStatus(Authentication auth) {
        boolean complete = profileService.isProfileComplete(auth.getName());
        return ResponseEntity.ok(Map.of("profileComplete", complete));
    }

    // ── Staff endpoints ───────────────────────────────────────────────────────

    @GetMapping("/api/admin/students")
    public ResponseEntity<Page<StudentProfileResponse>> getStudentList(
            Authentication auth,
            @RequestParam(required = false) Long blockId,
            @RequestParam(required = false) String roomNumber,
            @RequestParam(required = false) String name,
            @RequestParam(defaultValue = "0")  int page,
            @RequestParam(defaultValue = "20") int size) {
        return ResponseEntity.ok(
                profileService.getStudentList(auth.getName(), blockId, roomNumber, name, page, size));
    }

    @GetMapping("/api/admin/students/{userId}/profile")
    public ResponseEntity<StudentProfileResponse> getStudentProfile(
            @PathVariable Long userId, Authentication auth) {
        return ResponseEntity.ok(profileService.getStudentProfileByStaff(userId, auth.getName()));
    }

    @PutMapping("/api/admin/students/{userId}/profile")
    public ResponseEntity<StudentProfileResponse> updateStudentProfile(
            @PathVariable Long userId,
            @RequestBody StudentProfileUpdateRequest request,
            Authentication auth) {
        return ResponseEntity.ok(
                profileService.updateProfileByStaff(userId, request, auth.getName()));
    }

    @GetMapping("/api/admin/students/vacancies")
    public ResponseEntity<List<VacancyResponse>> getVacancies(
            @RequestParam Long blockId,
            Authentication auth) {
        return ResponseEntity.ok(profileService.getVacancyList(auth.getName(), blockId));
    }

    @GetMapping("/api/admin/students/export/csv")
    public ResponseEntity<byte[]> exportCsv(
            Authentication auth,
            @RequestParam(required = false) Long blockId,
            @RequestParam(required = false) String roomNumber,
            @RequestParam(required = false) String name) {
        byte[] csv = profileService.exportAsCsv(auth.getName(), blockId, roomNumber, name);
        return ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"students.csv\"")
                .contentType(MediaType.parseMediaType("text/csv"))
                .body(csv);
    }

    @GetMapping("/api/admin/students/export/pdf")
    public ResponseEntity<byte[]> exportPdf(
            Authentication auth,
            @RequestParam(required = false) Long blockId,
            @RequestParam(required = false) String roomNumber,
            @RequestParam(required = false) String name) {
        byte[] pdf = profileService.exportAsPdf(auth.getName(), blockId, roomNumber, name);
        return ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"students.pdf\"")
                .contentType(MediaType.APPLICATION_PDF)
                .body(pdf);
    }
}

File: backend/src/controllers/studentProfile.controller.js
const asyncHandler = require('../utils/asyncHandler');
const profileService = require('../services/studentProfile.service');

const submitProfile = asyncHandler(async (req, res) => {
  res.json(await profileService.submitProfile(req.user.email, req.body));
});

const getMyProfile = asyncHandler(async (req, res) => {
  res.json(await profileService.getMyProfile(req.user.email));
});

const checkProfileStatus = asyncHandler(async (req, res) => {
  const profileComplete = await profileService.isProfileComplete(req.user.email);
  res.json({ profileComplete });
});

// NOTE: `blockId` query param renamed to `block` ('A'/'B') — this app now
// serves one hostel, so blocks are no longer separate FK-addressable rows.
// Confirmed unused by the frontend before renaming. See
// docs/MIGRATION_NOTES.md "Hostel-10 scoping".
const getStudentList = asyncHandler(async (req, res) => {
  const { block, roomNumber, name } = req.query;
  const page = parseInt(req.query.page ?? '0', 10);
  const size = parseInt(req.query.size ?? '20', 10);
  res.json(await profileService.getStudentList(req.user.email, block, roomNumber, name, page, size));
});

const getStudentProfile = asyncHandler(async (req, res) => {
  res.json(await profileService.getStudentProfileByStaff(req.params.userId, req.user.email));
});

const updateStudentProfile = asyncHandler(async (req, res) => {
  res.json(await profileService.updateProfileByStaff(req.params.userId, req.body, req.user.email));
});

const getVacancies = asyncHandler(async (req, res) => {
  res.json(await profileService.getVacancyList(req.user.email, req.query.block));
});

const exportCsv = asyncHandler(async (req, res) => {
  const { block, roomNumber, name } = req.query;
  const csv = await profileService.exportAsCsv(req.user.email, block, roomNumber, name);
  res.setHeader('Content-Disposition', 'attachment; filename="students.csv"');
  res.setHeader('Content-Type', 'text/csv');
  res.send(csv);
});

const exportPdf = asyncHandler(async (req, res) => {
  const { block, roomNumber, name } = req.query;
  const pdf = await profileService.exportAsPdf(req.user.email, block, roomNumber, name);
  res.setHeader('Content-Disposition', 'attachment; filename="students.pdf"');
  res.setHeader('Content-Type', 'application/pdf');
  res.send(pdf);
});

module.exports = {
  submitProfile,
  getMyProfile,
  checkProfileStatus,
  getStudentList,
  getStudentProfile,
  updateStudentProfile,
  getVacancies,
  exportCsv,
  exportPdf,
};
